robot_smach_states/tutorial/challenge_tutorial.md

```diff
@@ -0,0 +1,236 @@
+* Now, we can add the following code in `example.py`:
+
+```python
+# ROS
+import smach
+
+# Robot smach states
+import robot_smach_states as states
+
+
+class Example(smach.StateMachine):
+    """ This is my example state machine
+
+    """
+    def __init__(self, robot):
+        """ Initialization method
+
+        :param robot: robot api object
+        """
+        smach.StateMachine.__init__(self, outcomes=["succeeded", "failed", "aborted"])
+
+        with self:
+            smach.StateMachine.add("HELLO_WORLD", states.Say(robot, "Hello world"), transitions={"spoken": "succeeded"})
+
+```
+* `import smach` imports the [smach](http://wiki.ros.org/smach) Python module so that we can use it
+* `import robot_smach_states as states` imports [the smach states](https://github.com/tue-robotics/tue_robocup/tree/master/robot_smach_states) that have been developed in the TU/e Robotics lab
+* `class Example(smach.StateMachine):` defines the statemachine that we are implementing. The `smach.StateMachine` between brackets shows that the `Example` class [inherits](https://en.wikipedia.org/wiki/Inheritance_(object-oriented_programming) from the `smach.StateMachine` class. (If you don't know what inheritance means you might want to study that first)
+* `def __init__(self, robot):` defines the signature of the initialization method. Basically, each class needs such a function. In this case, we need to supply the Robot object that was instantiated previously in our script.
+* `smach.StateMachine.__init__(self, outcomes=["succeeded", "failed", "aborted"])` calls the initialization method of the parent class. Here, we also define the outcomes of our statemachine to `succeeded`, `failed` and `aborted`.
+* `states.Say(robot, "Hello world")` instantiates a `Say` state using our robot object and the sentence "Hello world"
+* `smach.StateMachine.add` adds this state to our statemachine and calls it "HELLO_WORLD". Furthermore, it connects the one outcome of the `Say` state, being `spoken` to the `succeeded` outcome of our statemachine.
+
+### Adding our state machine to our app
+* Update `scripts/challenge_example` to:
+
+```python
+#!/usr/bin/python
+
+# ROS
+import rospy
+
+# TU/e Robotics
+from robot_skills.util.robot_constructor import robot_constructor
+
+# Challenge example
+from challenge_example.example import Example
+
+
+if __name__ == "__main__":
+
+    rospy.init_node("state_machine_test")
+    rospy.loginfo("Hello world")
+    robot = robot_constructor("amigo")
+    state_machine = Example(robot)
+    state_machine.execute()
+    raw_input("Press any key to exit")
+
+```
+* `from challenge_example.example import Example` imports the `Example` class that we defined in `src/challenge_example/example.py` into this scope.
+* `state_machine = Example(robot)` instantiates an Example statemachine using the robot object and calls it 'state_machine'.
+* `state_machine.execute()` executes the example state machine.
+
```

Comment:
Can not run example, give the error: ImportError: No module named challenge_example.example
This happens because no setup.py file is present, see link for the method to fix it: http://docs.ros.org/api/catkin/html/user_guide/setup_dot_py.html